csrc/transform_replay.cpp

```diff
@@ -233,6 +233,97 @@ TensorDomain* TransformReplay::fullSelfReplay(
       new_self_root->contiguity());
 }
 
+TensorDomain* TransformReplay::selfAllocationReplay(
+    const TensorDomain* new_self_root,
+    const TensorDomain* self) {
+  FUSER_PERF_SCOPE("TransformReplay::selfAllocationReplay");
+
+  // NOTE: We could also have reduction IDs involved in transformation that
+  // leads to allocation domain, so technically we should have included
+  // reduction IDs in the replay as well. The reason that we skipped them here
+  // is because this function is used by `RemoveBcastSqueeze`, where we could
+  // have mismatch reduction IDs on the logical between `self` and
+  // `new_self_root`.
+  auto new_self_logical = TensorDomain::noReductions(new_self_root->logical());
+  auto self_logical = TensorDomain::noReductions(self->logical());
+
+  NVF_ERROR(
+      new_self_logical.size() == self_logical.size(),
+      "Invalid number of IterDomains provided.");
+
+  // Map for replay
+  id_map axis_map;
+  {
+    int64_t i = 0;
+    for (auto id : self_logical) {
+      // Note: we don't want to check for equal `isRFactorProduct`, since we
+      // could replay Allocation of the output of a reduction to a later
+      // consumer tensor, which would not have the rfactor flag on.
+      NVF_ERROR(
+          new_self_logical[i]->isSymbolic() || id->isSymbolic() ||
+              new_self_logical[i]->isBroadcast() == id->isBroadcast(),
+          "Axes ",
+          id,
+          " and ",
+          new_self_logical[i],
+          " do not match for self replay.");
+      axis_map[id] = new_self_logical[i];
+      i++;
+    }
+  }
+
+  // Replay producer dimensions.
+  auto self_allocation = TensorDomain::noReductions(self->maybeAllocation());
+  ReplaySelf replay(self_allocation, axis_map);
+  std::vector<IterDomain*> mapped_alloc_domain(self_allocation.size(), nullptr);
+  std::vector<std::optional<bool>> mapped_contiguity = self->contiguity();
+
+  {
+    int64_t i = 0;
+    for (auto id : self_allocation) {
+      auto it = replay.getReplay().find(id);
+      NVF_ERROR(
+          it != replay.getReplay().end(),
+          "Error during replay, didn't replay an axis.");
+      if (it->second->isBroadcast() == mapped_contiguity[i].has_value()) {
+        // whether we resolve to true or false shouldn't matter since it's going
+        // to be concretized as a broadcast dimension
+        mapped_contiguity[i] =
+            it->second->isBroadcast() ? std::nullopt : std::make_optional(true);
+      }
+      mapped_alloc_domain[i++] = it->second;
+    }
+  }
+
+  // We need to put back the reduction IDs that are not mapped
```

Comment:
Yes but in a potentially different order. Is that fine? For example,
```
self: logical=[ix, iy, r], alloc=[iy, r, ix]
new_self: logical=[ix, iy, r]
```
will set new_alloc_domain to be `[iy, ix, r]` instead of `[iy, r, ix]`.

Reply:
I'm asking this because if the order of reductions doesn't matter we can simplify put reductions at the end.

Reply:
My general answer is that reduction *shouldn't* matter in allocation domain, where it might have some effect is that, we usually check for permutation from logical to allocation, and try to restore the order in loop domain for better memory access. Hence my attempt to keep reduction in the same place as in logical when possible.
But granted, that logic probably should have considered dropping reduction anyway, since it's pretty trivial.

TL;DR, I think you are right and I should just try to keep them at the end for simplicity.